File: app/src/main/java/no/nordicsemi/android/nrftoolbox/parser/GlucoseMeasurementContextParser.java
/*
 * Copyright (c) 2015, Nordic Semiconductor
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package no.nordicsemi.android.nrftoolbox.parser;

import no.nordicsemi.android.ble.data.Data;

@SuppressWarnings("ConstantConditions")
public class GlucoseMeasurementContextParser {
	private static final int UNIT_kg = 0;
	private static final int UNIT_l = 1;

	public static String parse(final Data data) {
		final StringBuilder builder = new StringBuilder();

		int offset = 0;
		final int flags = data.getIntValue(Data.FORMAT_UINT8, offset);
		offset += 1;

		final boolean carbohydratePresent = (flags & 0x01) > 0;
		final boolean mealPresent = (flags & 0x02) > 0;
		final boolean testerHealthPresent = (flags & 0x04) > 0;
		final boolean exercisePresent = (flags & 0x08) > 0;
		final boolean medicationPresent = (flags & 0x10) > 0;
		final int medicationUnit = (flags & 0x20) > 0 ? UNIT_l : UNIT_kg;
		final boolean hbA1cPresent = (flags & 0x40) > 0;
		final boolean moreFlagsPresent = (flags & 0x80) > 0;

		final int sequenceNumber = data.getIntValue(Data.FORMAT_UINT16, offset);
		offset += 2;

		if (moreFlagsPresent) // not supported yet
			offset += 1;

		builder.append("Sequence number: ").append(sequenceNumber);

		if (carbohydratePresent) {
			final int carbohydrateId = data.getIntValue(Data.FORMAT_UINT8, offset);
			final float carbohydrateUnits = data.getFloatValue(Data.FORMAT_SFLOAT, offset + 1);
			builder.append("\nCarbohydrate: ").append(getCarbohydrate(carbohydrateId)).append(" (").append(carbohydrateUnits).append(carbohydrateUnits == UNIT_kg ? "kg" : "l").append(")");
			offset += 3;
		}

		if (mealPresent) {
			final int meal = data.getIntValue(Data.FORMAT_UINT8, offset);
			builder.append("\nMeal: ").append(getMeal(meal));
			offset += 1;
		}

		if (testerHealthPresent) {
			final int testerHealth = data.getIntValue(Data.FORMAT_UINT8, offset);
			final int tester = (testerHealth & 0xF0) >> 4;
			final int health = (testerHealth & 0x0F);
			builder.append("\nTester: ").append(getTester(tester));
			builder.append("\nHealth: ").append(getHealth(health));
			offset += 1;
		}

		if (exercisePresent) {
			final int exerciseDuration = data.getIntValue(Data.FORMAT_UINT16, offset);
			final int exerciseIntensity = data.getIntValue(Data.FORMAT_UINT8, offset + 2);
			builder.append("\nExercise duration: ").append(exerciseDuration).append("s (intensity ").append(exerciseIntensity).append("%)");
			offset += 3;
		}

		if (medicationPresent) {
			final int medicationId = data.getIntValue(Data.FORMAT_UINT8, offset);
			final float medicationQuantity = data.getFloatValue(Data.FORMAT_SFLOAT, offset + 1);
			builder.append("\nMedication: ").append(getMedicationId(medicationId)).append(" (").append(medicationQuantity).append(medicationUnit == UNIT_kg ? "kg" : "l");
			offset += 3;
		}

		if (hbA1cPresent) {
			final float HbA1c = data.getFloatValue(Data.FORMAT_SFLOAT, offset);
			builder.append("\nHbA1c: ").append(HbA1c).append("%");
		}
		return builder.toString();
	}

	private static String getCarbohydrate(final int id) {
		switch (id) {
		case 1:
			return "Breakfast";
		case 2:
			return "Lunch";
		case 3:
			return "Dinner";
		case 4:
			return "Snack";
		case 5:
			return "Drink";
		case 6:
			return "Supper";
		case 7:
			return "Brunch";
		default:
			return "Reserved for future use (" + id + ")";
		}
	}

	private static String getMeal(final int id) {
		switch (id) {
		case 1:
			return "Preprandial (before meal)";
		case 2:
			return "Postprandial (after meal)";
		case 3:
			return "Fasting";
		case 4:
			return "Casual (snacks, drinks, etc.)";
		case 5:
			return "Bedtime";
		default:
			return "Reserved for future use (" + id + ")";
		}
	}

	private static String getTester(final int id) {
		switch (id) {
		case 1:
			return "Self";
		case 2:
			return "Health Care Professional";
		case 3:
			return "Lab test";
		case 4:
			return "Casual (snacks, drinks, etc.)";
		case 15:
			return "Tester value not available";
		default:
			return "Reserved for future use (" + id + ")";
		}
	}

	private static String getHealth(final int id) {
		switch (id) {
		case 1:
			return "Minor health issues";
		case 2:
			return "Major health issues";
		case 3:
			return "During menses";
		case 4:
			return "Under stress";
		case 5:
			return "No health issues";
		case 15:
			return "Health value not available";
		default:
			return "Reserved for future use (" + id + ")";
		}
	}

	private static String getMedicationId(final int id) {
		switch (id) {
		case 1:
			return "Rapid acting insulin";
		case 2:
			return "Short acting insulin";
		case 3:
			return "Intermediate acting insulin";
		case 4:
			return "Long acting insulin";
		case 5:
			return "Pre-mixed insulin";
		default:
			return "Reserved for future use (" + id + ")";
		}
	}
}

File: lib_profile/src/main/java/no/nordicsemi/android/toolbox/libs/profile/data/gls/GlucoseMeasurementContextParser.kt
package no.nordicsemi.android.toolbox.libs.profile.data.gls

import no.nordicsemi.android.kotlin.ble.core.data.util.DataByteArray
import no.nordicsemi.android.kotlin.ble.core.data.util.FloatFormat
import no.nordicsemi.android.kotlin.ble.core.data.util.IntFormat
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.Carbohydrate
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.GLSMeasurementContext
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.Health
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.Meal
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.Medication
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.MedicationUnit
import no.nordicsemi.android.toolbox.libs.profile.data.gls.data.Tester

object GlucoseMeasurementContextParser {

    fun parse(bytes: DataByteArray): GLSMeasurementContext? {

        if (bytes.size < 3) {
            return null
        }

        var offset = 0

        val flags: Int = bytes.getIntValue(IntFormat.FORMAT_UINT8, offset++) ?: return null
        val carbohydratePresent = flags and 0x01 != 0
        val mealPresent = flags and 0x02 != 0
        val testerHealthPresent = flags and 0x04 != 0
        val exercisePresent = flags and 0x08 != 0
        val medicationPresent = flags and 0x10 != 0
        val medicationUnitLiter = flags and 0x20 != 0
        val HbA1cPresent = flags and 0x40 != 0
        val extendedFlagsPresent = flags and 0x80 != 0

        if (bytes.size < (3 + (if (carbohydratePresent) 3 else 0) + (if (mealPresent) 1 else 0) + (if (testerHealthPresent) 1 else 0)
                    + (if (exercisePresent) 3 else 0) + (if (medicationPresent) 3 else 0) + (if (HbA1cPresent) 2 else 0)
                    + if (extendedFlagsPresent) 1 else 0)
        ) {
            return null
        }

        val sequenceNumber: Int = bytes.getIntValue(IntFormat.FORMAT_UINT16_LE, offset) ?: return null
        offset += 2

        // Optional fields
        if (extendedFlagsPresent) {
            // ignore extended flags
            offset += 1
        }

        var carbohydrate: Carbohydrate? = null
        var carbohydrateAmount: Float? = null
        if (carbohydratePresent) {
            val carbohydrateId: Int = bytes.getIntValue(IntFormat.FORMAT_UINT8, offset) ?: return null
            carbohydrate = Carbohydrate.create(carbohydrateId)
            carbohydrateAmount = bytes.getFloatValue(FloatFormat.FORMAT_SFLOAT, offset + 1) // in grams
            offset += 3
        }

        var meal: Meal? = null
        if (mealPresent) {
            val mealId: Int = bytes.getIntValue(IntFormat.FORMAT_UINT8, offset) ?: return null
            meal = Meal.create(mealId)
            offset += 1
        }

        var tester: Tester? = null
        var health: Health? = null
        if (testerHealthPresent) {
            val testerAndHealth: Int = bytes.getIntValue(IntFormat.FORMAT_UINT8, offset) ?: return null
            tester = Tester.create(testerAndHealth and 0x0F)
            health = Health.create(testerAndHealth shr 4)
            offset += 1
        }

        var exerciseDuration: Int? = null
        var exerciseIntensity: Int? = null
        if (exercisePresent) {
            exerciseDuration =
                bytes.getIntValue(IntFormat.FORMAT_UINT16_LE, offset) // in seconds
            exerciseIntensity =
                bytes.getIntValue(IntFormat.FORMAT_UINT8, offset + 2) // in percentage
            offset += 3
        }

        var medication: Medication? =
            null
        var medicationAmount: Float? = null
        var medicationUnit: MedicationUnit? = null
        if (medicationPresent) {
            val medicationId: Int = bytes.getIntValue(IntFormat.FORMAT_UINT8, offset) ?: return null
            medication = Medication.create(medicationId)
            medicationAmount = bytes.getFloatValue(FloatFormat.FORMAT_SFLOAT, offset + 1) // mg or ml
            medicationUnit =
                if (medicationUnitLiter) MedicationUnit.UNIT_ML else MedicationUnit.UNIT_MG
            offset += 3
        }

        var HbA1c: Float? = null
        if (HbA1cPresent) {
            HbA1c = bytes.getFloatValue(FloatFormat.FORMAT_SFLOAT, offset)
            // offset += 2;
        }

        return GLSMeasurementContext(
            sequenceNumber,
            carbohydrate,
            carbohydrateAmount,
            meal,
            tester,
            health,
            exerciseDuration,
            exerciseIntensity,
            medication,
            medicationAmount,
            medicationUnit,
            HbA1c
        )
    }
}